Switch to the FHRS API

urschrei opened the pull request:
This removes the raw legacy data and scraping logic, and switches to
using the UK FHRS API (see #6) instead. Requirements etc are also updated.

There are two drawbacks:
1. The metadata is less rich (i.e. there are no opening hours listed)
2. Not all entries (around 28 are missing from 391 total) in the raw data are geocoded. The processing data currently drops these records, but it could easily be adapted to Geocode them in a separate step using e.g. OpenCage, but it would mean users having to supply their own API key. Or we could provide two scripts: one that geocodes, and the standard one.

Also for discussion before merging:
I've also removed the legacy JSON output, since the retrieved metadata is less useful, and the retrieval script can easily be run in a Notebook if someone wants the full dataset. However, dumping a per-branch JSON entry is a quick change. Discuss!

urschrei committed:
Update to using the FHRS API

This removes the raw legacy data and scraping logic, and switches to
using the UK FHRS API instead.

urschrei committed:
Tidy up script

Jonty commented:
This is amazing 😍

So, this was Step One of what I planned to do, as I actually need the opening hours data for something I've been working on.

Step Two was using the postcodes from the FHRS data to hit up the Pret website and scrape the extended info for each Pret. I'm happy to implement this given you've done the hard bit, or if you're feeling excitable you can do it! Using the FHRS postcode _may_ also solve the geocoding problem, as Pret has the lat/lon hidden in the page data.

Step Three would be sorting out a circleci job that runs weekly to update the data, in [the same way I've done for other things](https://github.com/Jonty/uk_petitions_data/blob/master/.circleci/config.yml).

Thoughts?

urschrei commented:
Ooh. Well, the FHRS data API has very consistent post code data, so if the Pret scraping is reliable using post codes, maybe that's the way to go? Thus:

- Step 1: Get post codes for each Pret
- Step 2: Use post codes to scrape opening hours and lon / lat from Pret site
- Dump GeoJSON which now include opening hours (and interesting things such as FHRS ratings)

I've just had a look at the "find a Pret" page, and it looks pretty simple to scrape, so I may as well keep going with that. This rests on the assumption that the FHRS data is relatively current, though, and Google says there are 530 in the UK, vs the FHRS 398, which is annoying. Not sure what to do about that yet, though…

Jonty commented:
The alternative is to scrape every single postcode in the UK (or a subset, adjusted for density so you cover the "near" distance) - but that is just horrible.

I wonder if it's possible to get access to the food business registers: https://www.gov.uk/food-business-registration

_Some_ Prets are also [listed in local authority Licensing Act databases](https://www.cityoflondon.gov.uk/MVM/Online/EGov/License_Registers/StdResults.aspx?PT=Licensing%20Act%202003%20and%20Licensing%20Register&SC=Licence%20Type%20is%20LTC%20and%20Search%20Name%20is%20PRET&FT=Licence%20Registers%20Search%20Results&XMLSIDE=/MVM/SiteFiles/OnlineMenus/Licensing/Licensing.xml&XSLTemplate=/mvm/SiteFiles/Skins/M3PP_Online/xslt/Licensing/LicenceRegistersSearchResults.xsl&PS=10&XMLLoc=/MVM/Online/generic/XMLtemp/m4f4ql55kxfeaum1smz1oyy5/dc70029f-64ed-4582-ae4a-0d4872c05b72.xml), because they apply for "Chair and Table" licenses outside the property. I doubt Pret applies for this at every branch though.

urschrei committed:
Scrape opening hours from Pret website

NB: we're currently still dropping records for which we don't have
coordinates, though this will change once we have the geocoder running

urschrei commented:
> The alternative is to scrape every single postcode in the UK (or a subset, adjusted for density so you cover the "near" distance) - but that is just horrible.

I don't know whether we could run through all 1.7x million inside a CircleCI job before it timed out anyway, even if we parallelised it, although the idea does have a kind of frightening appeal.

I've got a fairly robust scraper working now (see latest commit), and I've written some geocoding logic, so we're getting there. I've also discovered that every "find my nearest Pret" result page has a LOT of "nearby" branches listed (e.g. W1T 4TJ gives me 319). I wonder whether it's just re-ordering the list geographically. Also, have a look at the latest GeoJSON, which now has opening hours data…does the format make sense?

urschrei committed:
Add geocoder

This assumes the availability of an OpenCage key in an env var named
OPENCAGE
